### ISSUE TITLE
S3DIS full-scenario inference

### ISSUE THREAD

**Issue by Buanma** [Matrix-ASC/DeLA_v2#2]
Hello, thank you for your wonderful work. During the process of replicating your work, I couldn't find the part of full-scenario reasoning. Is the LOOP in the test script similar to vote_num in point net++? (Sample multiple times until the entire scene is covered).

<img width="652" height="217" alt="Image" src="https://github.com/user-attachments/assets/fc7407df-da64-485a-a2ba-10ac89f822d7" />

<img width="484" height="140" alt="Image" src="https://github.com/user-attachments/assets/45e55b75-d83e-4027-ad02-b7999bb79552" />

 Also, could you please provide the relevant code for saving the prediction results (for example, saving them in xyz,pre format) for visualization purposes? Looking forward to your reply !

**Comment by binjiechen** [comment]
1. Yes. During test, a scene is grid subsampled and the predictions are upsampled and averaged (cum = cum + p[nn] where nn is nearest neighbor indices). Each time, hopefully a different point in a grid is selected (It's been too long and I have forgotten details by now). A point is not necessarily covered since preset vote num is  less than possible max point size in a grid. It runs faster with similar but no better performance.
2. In test script, cum is whole scene prediction logits. After line 41 (if cnt % loop == 0:), where a scene's prediction is done, you can save it. In get_test_item func of s3dis.py, you can return idx (defined here: idx //= self.loop), and get it in line 32 of test script (like for xyz, feature, indices, nn, y, idx in testdlr:), and save it along cum. So in a visualization script, you can load cum and idx, and load xyz and col as in s3dis script with idx. If you are not familiar with python/pytorch, you can ask an LLM to do it for you, just copy my descriptions here and the whole s3dis.py and test.py and ask it to adapt the code.

**Comment by Buanma** [comment]
Thank you for your reply. I have been urging the LLM to work for me, and the effect is quite good.
 I'd like to take this opportunity to ask for your advice again. I am currently training my own dataset (19,200/28,672 points per scene, with only two categories), and the sample size is extremely uneven (9:1). The accuracies I have obtained now are 92% and 85% (val iou) respectively, but this level of accuracy is still some distance away from practical application. 
At present, I have added the classweight of the loss function. Do you have any good suggestions?

I wonder if I could adjust the size of grid_size so that I can obtain more detailed information?
Looking forward to your guidance~

**Comment by binjiechen** [comment]
Applying DeLA v2 to a new dataset is not very straightforward. I'll try to cover all adaption details here.
1. Grid size. This should be set  so that points in one grid are mainly the same class. You can load a large representative scene, use indices = grid_subsampling(xyz, grid_size, 1.0) to get downsample indices, check  len(xyz) / len(indices)for downsampling ratio. You can start with an initial downsampling ratio close to 1. Then you set s3dis_args.grid_size = [init_grid_size, x2, x2, x2], and s3dis_warmup_args.grid_size = [init_grid_size, 3.5, 3.5, 3.5]. Normally, if you double grid_size, point size becomes 1/4. Warm up will train on the largest scene which helps pytorch pre-allocates memory. If you encounter unstable VRAM usage, set s3dis_warmup_args.grid_size = [init_grid_size, 3, 3, 3] or lower.  Note, the initial grid size must be properly set, so point cloud gets correctly downsampled. If initial grid size is too small, point gets no downsampling at all and the receptive field will be small. Oh, and set the third argument to grid_subsampling(_test) to 2.5 / downsampling ratio, or simply 1.
2. s3dis_args.max_pts = 30000. You can set this to an average scene point size after initial downsampling. You can also set it lower. This has an regularization effect. If it is large, network will always train on the whole scene and may overfit. But this is just my experience from S3DIS.
3. You may be using binary cross entropy. You can try cross entropy with label smoothing.  label smoothing is useful for label imbalance.  If you use classweight, you can set set weight to 1:3 instead of 1:9. Or 1 : 9**(1/3). Just experience, might not work.
4. For the network and relative coordinate prediction regularization to work properly. dela_args.cor_std and all_dist must be reset. First, set cor_std to all 1, then uncomment line 225 (# print(rel_cor.std(dim=0))), and run then training for some iterations, check output log, cor_std is printed multiple times for each layer, set the values. The order is reversed, anyway the smallest one is for the 1st layer. After you set it, comment 225. Then uncomment 175 to 185, run train.py, it will self exit. The folder should have dist0.pt to dist3.pt. Comment 175 to 185. Ok now.
5. Data augmentation. There are multiple data augmentations. It's best to test all of them out. For example, rotmat *= random.uniform(0.8, 1.2) for random scaling, xyz += torch.empty_like(xyz).normal_(std=0.005) for random jittering. Turn on/off or adjust magnitude. Cutout is also important. You can also try mix3D as in ScanNetv2 setting. But that's a little more complicated and I have forgotten details.
6. In line 118 to 126 of s3dis.py, inputs are normalized, for example, color / 250, height / 3.3, xyz * 40 (as relative xyz is used). These should just work fine if the color input is 0-255, scene max height is around 3. If you want to adjust, just scale height should be fine, because xyz is rescaled by cor_std inside network. Don't forget to adjust get_test_item if you edit here.
7. Based on how large your dataset is, tune network dim and depth, and training epoches, batch_size and learning rate. Learning rate, batch size and training epoches are very important! In S3DIS setting, a dataset is looped 30 times, so actual epoches is  epoch x 30. Set warmup to 1/10 of epoch.  Tuning these hyper-parameters is more a thing of experience, so you achieve a good balance between underfitting and overfitting.
8. Maybe something else, but I can't recall any right now. So that's all.

**Comment by Buanma** [comment]
Thank you for your patient and detailed reply. These suggestions are very helpful to me. 

And I'm very sorry to disturb you again. I often encounter memory errors when loading data (but S3DIS doesn't). Like“malloc_consolidate(): invalid chunk size” or "free(): invalid next size (fast)". After checking, I found that it seems to be a problem with **grid_subsampling** in the C++ extension, but I'm not quite sure exactly why the problem occurs, and the LLM has not solved it either. I pasted the modified **_ _getitem_ _** method and the error log as follows.
I would be most grateful if you could help me check this small bug again in your spare time. 

Wish you a smooth life and study！

<img width="1480" height="790" alt="Image" src="https://github.com/user-attachments/assets/b828bdc3-6411-486a-98f0-fa79c4092c49" />

<img width="1803" height="791" alt="Image" src="https://github.com/user-attachments/assets/02c09c37-a24d-4854-9afe-bd5e632f26f2" />

    def __getitem__(self, idx):
        idx = int(idx)
        if self.test:
            return self.get_test_item(idx)

        idx //= self.loop
        xyz, lbl = self.datas[idx]
        # print(idx)

        if self.train:
            angle = random.random() * 2 * math.pi
            cos, sin = math.cos(angle), math.sin(angle)
            rotmat = torch.tensor([[cos, sin, 0], [-sin, cos, 0], [0, 0, 1]])
            rotmat *= random.uniform(0.8, 1.2)
            xyz = xyz @ rotmat
            xyz += torch.empty_like(xyz).normal_(std=0.005)
            xyz -= xyz.min(dim=0)[0]

        # here grid size is assumed 0.04, so estimated downsampling ratio is ~14
        if self.train:
            indices = grid_subsampling(xyz, self.grid_size[0], 1.0)
        else:
            indices = grid_subsampling_test(xyz, self.grid_size[0], 1.0, pick=0)

        xyz = xyz[indices]

        if not self.train:
            xyz -= xyz.min(dim=0)[0]

        if xyz.shape[0] > self.max_pts and self.train:
            pt = random.choice(xyz)
            condition = (xyz - pt).square().sum(dim=1).argsort()[:self.max_pts].sort()[0]  # sort to preserve locality
            xyz = xyz[condition]
            indices = indices[condition]

        lbl = lbl[indices]


        # normalized xyz replace color info
        xyz_norm = (xyz - xyz.mean(dim=0, keepdim=True)) / (xyz.std(dim=0, unbiased=False, keepdim=True) + 1e-6)

        height = xyz[:, 2:] * (1 / 3.3)
        feature = torch.cat([xyz_norm, height], dim=1)

        indices = []
        self.knn(xyz, self.grid_size[::-1], self.k[::-1], indices)

        xyz.mul_(40)

        return xyz, feature, indices, lbl

**Comment by binjiechen** [comment]
According to this code snippet, I can't see a problem.
This error can happen when the third argument (the size of pre-allocated hash table) to grid_subsampling is too small, but theoretically this should not happen when it is 1.0 (1 slot for 1 point).
I suggest set the third argument from 1.0 to 2.0 to all calls of grid_subsampling(_test), including the one in function knn.
I also suggest double checking all the modifications you made compared to s3dis version.
There must be a reason, but according to what I currently know, I can't determine it.

**Comment by Buanma** [comment]
Thank you so much! It worked. I adjusted all the sizes to 8.
Then I asked the LLM, and the reply was: "1.0 is sufficient" is based on the ideal assumption of "no conflict, no loss, and perfect hash distribution", but the density of actual point cloud data, the limitations of the hash function, and the implicit computation of the code will all lead to insufficient slots. However, 8.0, by "over-allocating slots", traded space for stability, covering the problems in these actual scenarios, and thus no longer reported errors.
Based on the above explanation, there are indeed many duplicate points (the origin of the camera coordinate system) in my corresponding data, which should be the factor causing hash collisions!
Thank you for your patient guidance,  Best wishes to you!

**Comment by binjiechen** [comment]
I just found the bug in my code. It has nothing to do with hash collision. It's just that there is one reserved slot, so minimum slot size should be point size + 1. This bug only occurs when subsampled point size == original point size and that the size is a prime.
This suggests you may be using a grid size that is too small. This is usually not good. For the initial input, grid subsampling can be beneficial to generate a more uniform point cloud. For later stages, network needs a sparser point cloud. In either case, grid size can be larger.
In your screenshot, error occurs at the second call to knn (second downsampling stage, grid_size[2]). Or maybe your first grid sizes are too large that the point cloud is already very sparse? I can't tell. 
Back to the bug, to fix it, in utils/cutils/\_\_init__.py line 46, simply replace `size = next_prime(size)` with `size = next_prime(size + 1)`. And you can set the third argument back to 1.
Also, never trust your LLM too easily.

**Comment by Buanma** [comment]
I tried your suggestion and it was indeed like this. Thank you very much for your help
